source/application/AMPIC887DriverServer/AMPIC887AxisMap.h

```diff
@@ -0,0 +1,297 @@
+#ifndef AMPIC887AXISMAP_H
+#define AMPIC887AXISMAP_H
+
+#include <QMap>
+#include <QList>
+#include "AMGCS2.h"
+#include "AMPIC887AxisCollection.h"
+#include "AMGCS2Support.h"
+#include "util/AMCArrayHandler.h"
+/*!
+  * A class which maps Axes of a PI C887.11 controller to a type T.
+  */
+template <typename T>
+class AMPIC887AxisMap : public QMap<AMGCS2::Axis, T>
+{
+public:
+	/*!
+	  * Creates an instance of an axis map, which is empty.
+	  */
+	AMPIC887AxisMap() {}
+
+	/*!
+	  * Returns a collection of the axes which this map contains.
+	  */
+	AMPIC887AxisCollection axes() const
+	{
+		AMPIC887AxisCollection returnCollection(AMPIC887AxisCollection::EmptyCollection);
+		QList<AMGCS2::Axis> axisEntries = this->keys();
+
+		foreach(AMGCS2::Axis currentAxisEntry, axisEntries) {
+			returnCollection.append(currentAxisEntry);
+		}
+
+		return returnCollection;
+	}
+
+	/*!
+	  * Outputs the map values as a human readable string, if it can parse the
+	  * type into a string, otherwise returns the empty string.
+	  */
+	QString toString(const QString& = "Data") const
+	{
+		// Implemented for known types in partial specializations below.
+		return QString();
+	}
+
+	/*!
+	  * Whether this map contains an entry for the UnknownAxis.
+	  */
+	bool containsUnknownAxis() const
+	{
+		return this->contains(AMGCS2::UnknownAxis);
+	}
+
+	/*!
+	  * A map of the provided axes to their values stored within this map. If
+	  * a key contained within the provided collection is not contained within
+	  * this map it will not be included in the returned map.
+	  * \param keys ~ A collection of axes whose values within this map will
+	  * be returned.
+	  */
+	AMPIC887AxisMap<T> values(const AMPIC887AxisCollection& keys) const
+	{
+		AMPIC887AxisMap<T> returnValues;
+		foreach(AMGCS2::Axis currentAxis, keys) {
+			if(this->contains(currentAxis)) {
+				returnValues.insert(currentAxis, this->value(currentAxis));
+			}
+		}
+
+		return returnValues;
+	}
+
+};
+
+// Partial Specialization on type T to provide functionality of toString() method
+////////////////////////////////////////////////////////////////////////////////
+
+template <>
+class AMPIC887AxisMap<bool> : public QMap<AMGCS2::Axis, bool>
+{
+public:
+	/*!
+	  * Creates an instance of an axis map, which is empty.
+	  */
+	AMPIC887AxisMap() {}
+
+	/*!
+	  * Returns a collection of the axes which this map contains.
+	  */
+	AMPIC887AxisCollection axes() const
+	{
+		AMPIC887AxisCollection returnCollection(AMPIC887AxisCollection::EmptyCollection);
```

Comment:
Same comment here.  I'll stop repeating myself if I find other instances.  However, how often do you call `axes()`?  If it's really often, could this be done in such a way that you don't have to generate a fresh list every time you called it?  I could image a situation where you have to repeatedly call this function which would be much faster as some sort of cached set of axes.  Comments?